src/jdk.compiler/share/classes/com/sun/tools/javac/code/Lint.java

```diff
@@ -429,19 +428,27 @@ private EnumSet<LintCategory> suppressionsFrom(JCAnnotation annotation) {
         return suppressionsFrom(Stream.of(annotation).map(anno -> anno.attribute));
     }
 
-    // Find the @SuppressWarnings annotation in the attribute stream and extract the suppressions
+    // Find the @SuppressWarnings annotation in the given stream and extract the recognized suppressions
     private EnumSet<LintCategory> suppressionsFrom(Stream<Attribute.Compound> attributes) {
         initializeIfNeeded();
-        return attributes
+        EnumSet<LintCategory> result = LintCategory.newEmptySet();
+        attributes
           .filter(attribute -> attribute.type.tsym == syms.suppressWarningsType.tsym)
-          .map(attribute -> attribute.member(names.value))
-          .flatMap(attribute -> Stream.of(((Attribute.Array)attribute).values))
-          .map(Attribute.Constant.class::cast)
-          .map(elem -> elem.value)
-          .map(String.class::cast)
-          .map(LintCategory::get)
-          .filter(Objects::nonNull)
-          .collect(Collectors.toCollection(LintCategory::newEmptySet));
+          .map(this::suppressionsFrom)
+          .forEach(result::addAll);
+        return result;
+    }
+
+    // Given a @SuppressWarnings annotation, extract the recognized suppressions
+    private EnumSet<LintCategory> suppressionsFrom(Attribute.Compound suppressWarnings) {
+        EnumSet<LintCategory> result = LintCategory.newEmptySet();
+        Attribute.Array values = (Attribute.Array)suppressWarnings.member(names.value);
+        for (Attribute value : values.values) {
```

Comment:
I think this loop is more stream-friendly - e.g. you start from `values.values` then:
* `flatMap` using `Optional.of(value).stream()`
* `map` (using `LintCategory::get`, as now)
* `collect` to a new enum set

Reply:
I used a loop based on the "use imperative logic" suggestion, but maybe that's not what you meant... In any case I agree it's somehow not fully satisfying, as there's no ideal solution here because of all the casting and null checking required.